File: VB/App_Code/DataProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq
Imports System.Web

''' <summary>
''' Summary description for DataProvider
''' </summary>
Public Class DataProvider
    Public Shared Property DataSessionKey() As String
        Get
            Dim value As Object = HttpContext.Current.Session("DataSessionKey")
            Return If(value IsNot Nothing, DirectCast(value, String), "CategoryProduct")
        End Get
        Set(ByVal value As String)
            HttpContext.Current.Session("DataSessionKey") = value
        End Set
    End Property

    Public Shared Function GetMasterData() As DataTable
        Dim data As DataSet = GetData()
        If data IsNot Nothing Then
            Return data.Tables("Master")
        Else
            Return Nothing
        End If
    End Function

    Public Shared Function GetDetailData(ByVal masterRowKey As Object) As DataView
        Dim data As DataSet = GetData()
        If data IsNot Nothing Then
            Dim detail As DataTable = data.Tables("Detail")
            Dim columnName As String = data.Relations("MasterDetail").ParentColumns(0).ColumnName
            Return New DataView(detail, String.Format("[{0}] = '{1}'", columnName, masterRowKey), String.Empty, DataViewRowState.CurrentRows)
        Else
            Return Nothing
        End If
    End Function

    Private Shared Function GetData() As DataSet
        Select Case DataSessionKey
            Case "CategoryProduct"
                Return GetCategoryProductData()
            Case "CustomerOrder"
                Return GetCustomerOrderData()
            Case Else
                Return Nothing
        End Select
    End Function

    Private Shared Function GetCategoryProductData() As DataSet
        Dim data As DataSet = DirectCast(HttpContext.Current.Session("CategoryProduct"), DataSet)
        If data Is Nothing Then
            data = RequestData("SELECT [CategoryID], [CategoryName], [Description] FROM [Categories]", "CategoryID", "SELECT [ProductID], [ProductName], [UnitPrice], [Discontinued], [CategoryID] FROM [Products]", "ProductID")
            HttpContext.Current.Session("CategoryProduct") = data
        End If
        Return data
    End Function

    Private Shared Function GetCustomerOrderData() As DataSet
        Dim data As DataSet = DirectCast(HttpContext.Current.Session("CustomerOrder"), DataSet)
        If data Is Nothing Then
            data = RequestData("SELECT [CustomerID], [CompanyName], [ContactTitle], [City], [Country] FROM [Customers]", "CustomerID", "SELECT [OrderID], [OrderDate], [ShipName], [ShipCity], [ShipCountry], [CustomerID] FROM [Orders]", "OrderID")

            HttpContext.Current.Session("CustomerOrder") = data
        End If
        Return data
    End Function

    Private Shared Function RequestData(ByVal masterSelectCommand As String, ByVal masterTableKeyField As String, ByVal detailSelectCommand As String, ByVal detailTableKeyField As String) As DataSet
        Dim masterData As DataTable = DataHelper.ProcessSelectCommand(masterSelectCommand)
        masterData.PrimaryKey = New DataColumn() { masterData.Columns(masterTableKeyField) }
        masterData.TableName = "Master"

        Dim detailData As DataTable = DataHelper.ProcessSelectCommand(detailSelectCommand)
        detailData.PrimaryKey = New DataColumn() { detailData.Columns(detailTableKeyField) }
        detailData.TableName = "Detail"

        Dim data As New DataSet()
        data.Tables.Add(masterData)
        data.Tables.Add(detailData)
        data.Relations.Add("MasterDetail", masterData.Columns(masterTableKeyField), detailData.Columns(masterTableKeyField))

        Return data
    End Function
End Class

File: CS/App_Code/DataProvider.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Web;

/// <summary>
/// Summary description for DataProvider
/// </summary>
public class DataProvider
{
	public static string DataSessionKey {
		get {
			object value = HttpContext.Current.Session["DataSessionKey"];
			return (value != null) ? (string)value : "CategoryProduct";
		}
		set {
			HttpContext.Current.Session["DataSessionKey"] = value;
		}
	}

	public static DataTable GetMasterData() {
		DataSet data = GetData();
		if (data != null)
			return data.Tables["Master"];
		else
			return null;
	}

	public static DataView GetDetailData(object masterRowKey) {
		DataSet data = GetData();
		if (data != null) {
			DataTable detail = data.Tables["Detail"];
			string columnName = data.Relations["MasterDetail"].ParentColumns[0].ColumnName;
			return new DataView(detail, String.Format("[{0}] = '{1}'", columnName, masterRowKey), String.Empty, DataViewRowState.CurrentRows);
		}
		else
			return null;
	}

	private static DataSet GetData() {
		switch (DataSessionKey) {
			case "CategoryProduct":
				return GetCategoryProductData();
			case "CustomerOrder":
				return GetCustomerOrderData();
			default:
				return null;
		}
	}

	private static DataSet GetCategoryProductData() {
		DataSet data = (DataSet)HttpContext.Current.Session["CategoryProduct"];
		if (data == null) {
			data = RequestData("SELECT [CategoryID], [CategoryName], [Description] FROM [Categories]"
				, "CategoryID"
				, "SELECT [ProductID], [ProductName], [UnitPrice], [Discontinued], [CategoryID] FROM [Products]"
				, "ProductID");
			HttpContext.Current.Session["CategoryProduct"] = data;
		}
		return data;
	}

	private static DataSet GetCustomerOrderData() {
		DataSet data = (DataSet)HttpContext.Current.Session["CustomerOrder"];
		if (data == null) {
			data = RequestData("SELECT [CustomerID], [CompanyName], [ContactTitle], [City], [Country] FROM [Customers]"
				, "CustomerID"
				, "SELECT [OrderID], [OrderDate], [ShipName], [ShipCity], [ShipCountry], [CustomerID] FROM [Orders]"
				, "OrderID");

			HttpContext.Current.Session["CustomerOrder"] = data;
		}
		return data;
	}

	private static DataSet RequestData(string masterSelectCommand, string masterTableKeyField, string detailSelectCommand, string detailTableKeyField) {
		DataTable masterData = DataHelper.ProcessSelectCommand(masterSelectCommand);
		masterData.PrimaryKey = new DataColumn[] { masterData.Columns[masterTableKeyField] };
		masterData.TableName = "Master";

		DataTable detailData = DataHelper.ProcessSelectCommand(detailSelectCommand);
		detailData.PrimaryKey = new DataColumn[] { detailData.Columns[detailTableKeyField] };
		detailData.TableName = "Detail";

		DataSet data = new DataSet();
		data.Tables.Add(masterData);
		data.Tables.Add(detailData);
		data.Relations.Add("MasterDetail", masterData.Columns[masterTableKeyField], detailData.Columns[masterTableKeyField]);

		return data;
	}
}